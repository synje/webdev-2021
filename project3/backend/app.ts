import express from 'express';
import { graphqlHTTP } from 'express-graphql'
import mongoose from 'mongoose'

import rootScheme from './rootSchema'
import rootResolver from './rootResolver';
require('dotenv').config({path: './db.env'})

/**
 * Set up app with express
 */
const app = express();

/**
 * Sets headers for the app call to html. This is to avoid getting stopped by CORS
 */
app.use((req: any, res: any, next: any) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST,GET,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  if (req.method === 'OPTIONS') {
     return res.sendStatus(200);
  }
 next();
 });

 /**
  * Sets up the graphQL backend with schema and resolver
  */
app.use('/graphql', graphqlHTTP({
  schema: rootScheme,
  rootValue: rootResolver,
  graphiql: true
}))

/**
 * Connects to MongoDB database on VM
 */
mongoose.connect(`mongodb://it2810:gjest3-30@it2810-30.idi.ntnu.no:27017/it2810?authSource=it2810`).then(() => {
  app.listen(3002, "0.0.0.0", )
  console.log('Connected')
}).catch(err => {
  console.log(err);
  throw err
})
