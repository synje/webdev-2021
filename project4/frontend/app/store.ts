import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { isNotEmittedStatement } from 'typescript';
import termsReducer from './termsSlice';

export const store = configureStore({
  reducer: {
    terms: termsReducer
  },
});


export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
