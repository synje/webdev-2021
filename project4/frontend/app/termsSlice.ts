import React from 'react'
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from './store';

export interface TermsState {
    searchTerm: string;
    sortTerm: string;
    status: 'idle' | 'loading' | 'failed';
  }

const initialState: TermsState = {
    searchTerm: "",
    sortTerm: "",
    status: 'idle',
  };

  /**
   * Make reducers for searcTerm and sortTerm
   */
export const termsSlice = createSlice({
    name: 'terms',
    initialState,
    reducers:  {
        updateSearchTerm : (state, action: PayloadAction<string>) => {
            state.searchTerm = action.payload
        },
        updateSortTerm : (state, action: PayloadAction<string>) => {
            state.sortTerm = action.payload
        }
    }
})

export const {updateSortTerm, updateSearchTerm} = termsSlice.actions;
export const selectSearchTerm = (state: RootState) => state.terms.searchTerm
export const selectSortTerm = (state: RootState) => state.terms.sortTerm

export default termsSlice.reducer