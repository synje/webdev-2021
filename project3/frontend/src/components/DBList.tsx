import { useState } from 'react'
import SearchBar from './SearchBar'
import { List, ListItem } from '@chakra-ui/layout'
import { useQuery, gql } from "@apollo/client";
import { useAppSelector, useAppDispatch } from '../app/hooks';
import { updateSortTerm, updateSearchTerm, selectSearchTerm, selectSortTerm } from '../app/termsSlice';
import { Text, useDisclosure, Button, Modal, Box, Heading } from "@chakra-ui/react"
import ViewSong from './ViewSong';

/**
 * Query structure for querey to backend
 * Used in useQuyery call in DBList component
 */
 export const GET_FILTERED_SONGS = gql`
    query GetFilterSongs(
        $filter: String!
        $sortField: String
        $limit: Int
        $offset: Int
        ) {
        filterSongs(filter: $filter, sortField: $sortField, limit: $limit, offset: $offset){
            _id
            title
            duration
            album
            year_released
            artist{
                name1
                name2
            }
        }}
    `

/**
 * Exports DBList component (everything on the page under the red line) to App.tsx
 */
const DBList = () => {

    /** Functionality for: VIEWSONG
     * State for opening and closing ViewSong-window (Modal from Chakra)
     * Stores data from list and database to property (sends property to ViewSong component further down the file)
     */
    const { isOpen, onOpen, onClose } = useDisclosure()
    const [property, setProperty] = useState({ _id: "", title: "", duration: 0, album: "", year: 0, artist: [""] })
    function getView(_id: string, title: string, duration: number, album: string, year: number, artist: Array<any>) {
        let artistArray: any[] = []
        artist.map((name: any) => {
            artistArray.push(name.name1)
            if (artist[0].name2) {
                artistArray.push(", ", name.name2)
            }
        })
        setProperty({ _id: _id, title: title, duration: duration, album: album, year: year, artist: artistArray })
    }
    function onOpenFunc() { return onOpen() }
    function onCloseFunc() { return onClose() }


    /** Functionality for: SEARCH/FILTER AND LISTVIEW */
    const dispatch = useAppDispatch();
    /**
     * Variables for searchTerm. SearchTerm is used for saving values in searchbar onChange, and fetching query at load, while prevSearchTerm is used for fetching more queries.
     * PrevSearchTerm uses redux state searchTerm
     */
    let searchTerm = ""
    let prevSearchTerm2 = useAppSelector(selectSearchTerm);
    /**
     * Variables for sortTerm. SorTterm is used for saving values from chosen sort, and fetching query at load, while prevSortTerm is used for fetching more queries.
     * PrevSearchTerm uses redux state searchTerm
     */
    let sortTerm = useAppSelector(selectSortTerm)
    let savedSortTerm2 = useAppSelector(selectSortTerm);

    let count = 0 // Counter for displaying number of elements currently loaded on page
    /**
     * Fetching the data from the query. 
     * Takes in parameters for filter, with field to sort by, the limit of how many elements should be fetched and the offset of which elements are fetched
     * Loading is displayed at the start of the fething of data, and if there is an error in the fetch, error is displayed
     */
    const { error, loading, data, fetchMore } = useQuery(GET_FILTERED_SONGS, {
        variables: {
            filter: searchTerm,
            sortField: sortTerm,
            limit: 10,
            offset: 0
        }
    });
    if (loading) return <p>Loading...</p>;
    if (error) return <p>Error :(</p>;


    // GET TITLE
    /**
     * Slices the title if its too long and adds ...
     * @param title of song as string
     * @returns title as string
     */
    function getTitleString(title: string) {
        if (title.length > 10) {
            title = title.slice(0, 10)
            return title + "..."
        }
        return title
    }
    // GET ARTIST
    /**
     * Slices the artist-names if its too long and adds ...
     * @param artist1 of song as string
     * @param artist2 of song as any (can be undefined or null, because it's not required)
     * @returns 
     */
    function getArtistString(artist1: String, artist2: any) {
        if (artist1.length > 10) {
            artist1 = artist1.slice(0, 10) + "..."
        }
        if (typeof (artist2) === "string") {
            if (artist2.length > 5) {
                artist2 = artist2.slice(0, 5) + "..."
            }
            let artists: String = artist1 + ", " + artist2
            return artists
        }
        return artist1
    }
    // ADDS SECOND ARTIST TO LISTVIEW IF PRESENT
    /**
     * Returns artist2 as string if there is input from the user in this field
     * @param artist2 of song as any (can be undefined or null, because it's not required)
     * @returns artist2 as string
     */
    function getDuo(artist2: any) {
        if (typeof (artist2) === "string") {
            return artist2
        }
    }


    // S T Y L I N G :
    /**
     * Styling-variables to different Chakra(HTML)-elements on the page
     */
    // STYLING FOR TEXT IN LISTVIEW
    const textStyle = {
        color: "white",
        fontSize: "17px",
        fontweight: "bold"
    }
    // STYLING FOR LISTITEMS IN LISTVIEW
    const listItemStyle = {
        backgroundColor: "darkred"
    }
    // STYLING FOR BUTTONS ON LISTVIEW
    const viewButtonStyle = {
        backgroundColor: "#a23333",
        color: "white",
        border: "0.2px solid white"
    }
    // STYLING FOR BUTTONS ON SHOW MORE/LESS
    const buttonStyle = {
        backgroundColor: "floralwhite",
        border: "0.2px solid dimgray"
    }
    // STYLING FOR BUTTONS ON SORT
    const sortButtonStyle = {
        backgroundColor: "floralwhite",
        border: "0.2px solid darkred",
        borderRadius: "0",
        marginRight: "2px",
        marginBottom: "2px"
    }



    // SORTING
    /**
     * Function for fetching more data, with the previou searchTerm (the one saved),
     * a limit with is determined by the lenght of the current amout of displeyed data and the current sortTerm
     * @returns data, from a new query with new variables
     */
    function fetchingMoreDataSort(){
        dispatch(updateSortTerm(sortTerm))
        let offSetDataLength = parseInt(data.filterSongs.length)
        return fetchMore({
            variables: { filter: prevSearchTerm2, limit: offSetDataLength, sortField: sortTerm },
            updateQuery: (prevResult, { fetchMoreResult }) => {
                return fetchMoreResult;
    
            }
        })
    }

    return (
        <>
            {/* HEADER */}
            <Box w="100%" mt="3%" mb="3%" display="flex" alignItems="center" flexDirection="column">
                <Heading display="flex" justifyContent="flex-start" w="90%">Search the collection:</Heading>
            </Box>

            {/* FOR VIEWSONG */}
            <Modal data-cy= "songBox" onClose={onCloseFunc} isOpen={isOpen} isCentered>
                <ViewSong param={property} />
            </Modal>

            {/* SEARCHBAR */}
            <Box w="100%">
                <Box display="flex" justifyContent="center" w="100%"><Box as ="span" w="90%" fontSize="110%">Search for songs, albums or artists</Box></Box>
                <SearchBar
                    searchInput={(e: any) => searchTerm = e.target.value}
                    placeholder="Search..."
                    handleClick={(e: any) => {
                        e.preventDefault()
                        dispatch(updateSearchTerm(searchTerm))
                        fetchMore({
                            variables: { filter: searchTerm, sortField: sortTerm },
                            updateQuery: (prevResult, { fetchMoreResult }) => {
                                return fetchMoreResult;
                            }
                        })
                    }}
                />
                <Box className="sortingNav" mb="3%" mt="3%" display="flex" justifyContent="center">
                    <Box w="90%">
                        <Button w="240px" data-cy="sortbyTitle" style={sortButtonStyle} onClick={() => { sortTerm = 'title', fetchingMoreDataSort() }}>Sort by title (ASC)</Button>
                        <Button w="240px" style={sortButtonStyle} onClick={() => { sortTerm = "duration", fetchingMoreDataSort() }}>Sort by duration (ASC)</Button>
                        <Button w="240px" style={sortButtonStyle} onClick={() => { sortTerm = "album", fetchingMoreDataSort() }}>Sort by album (ASC)</Button>
                        <Button w="240px" style={sortButtonStyle} onClick={() => { sortTerm = "year_released", fetchingMoreDataSort() }}>Sort by year released (ASC)</Button>
                        <Button w="240px" style={sortButtonStyle} onClick={() => { sortTerm = "artist", fetchingMoreDataSort() }}>Sort by artist (ASC)</Button>
                        <Button w="240px" data-cy="removeSort" style={sortButtonStyle} onClick={() => { sortTerm = "", fetchingMoreDataSort() }}>Remove sort</Button>
                    </Box>
                </Box>
            </Box>

            {/* LISTVIEW */}
            <Box display="flex" justifyContent="center" w="100%">
                <List spacing={2} w="90%">
                    {data.filterSongs.map((song: { _id: string; title: string; duration: number; album: string; year_released: number, artist: Array<any> }) => (
                        <>
                            <ListItem  data-cy= "listItem" key={count + 1} style={listItemStyle} className="listItems" display="flex" justifyContent="space-around" alignItems="center" bg="tomato" w="100%" p={4} color="white" borderRadius={3}>
                                {" "}
                                <Text mr="15px" style={textStyle} id="songCount" className="listText">{count += 1}</Text>
                                <Text mr="13px" style={textStyle} data-cy="songTitle" id="songTitle" className="listText">{getTitleString(song.title)}</Text>
                                {song.artist.map((artist: { name1: String; name2: String | null }) => (
                                    <>
                                        <Text mr="13px" style={textStyle} className="listText" id="songArtist">
                                            {getArtistString(artist.name1, getDuo(song.artist[0].name2))}
                                        </Text>
                                    </>
                                ))}

                                {/* FOR VIEW SONG */}
                                <Button style={viewButtonStyle} onClick={() => {
                                    onOpenFunc()
                                    getView(song._id, song.title, song.duration, song.album, song.year_released, song.artist)
                                }
                                } data-cy= "songView" color="darkslategrey">VIEW SONG</Button>
                            </ListItem>
                        </>
                    ))}
                </List>
            </Box>

            {/* SHOW MORE/LESS BUTTONS */}
            <Box display="flex" justifyContent="center" mt="3%" mb="3%" w="100%">
                <Button data-cy="showMore" mr="10px" style={buttonStyle} onClick={() => {
                    let offSetDataLength = parseInt(data.filterSongs.length)
                    fetchMore({
                        variables: { filter: prevSearchTerm2, offset: offSetDataLength, sortField: savedSortTerm2 },
                        updateQuery: (prevResult, { fetchMoreResult }) => {
                            fetchMoreResult.filterSongs = [
                                ...prevResult.filterSongs,
                                ...fetchMoreResult.filterSongs
                            ];
                            return fetchMoreResult;

                        }
                    })
                }} >Show more</Button>
                <Button data-cy= "showLess" style={buttonStyle} onClick={() => {
                    fetchMore({
                        variables: { filter: prevSearchTerm2, offset: 0, limit: 10, sortField: savedSortTerm2 },
                        updateQuery: (prevResult, { fetchMoreResult }) => {
                            return fetchMoreResult;

                        }
                    })
                }} >Show less</Button>
            </Box>
        </>
    )
}

export default DBList;