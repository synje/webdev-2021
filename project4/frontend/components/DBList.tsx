import React, { useState } from 'react'
import { useQuery, gql } from "@apollo/client"
import { useAppSelector, useAppDispatch } from '../app/hooks'
import { updateSortTerm, updateSearchTerm, selectSearchTerm, selectSortTerm } from '../app/termsSlice'
import { View, Text, Platform, StyleSheet, Dimensions, ScrollView, TouchableOpacity } from 'react-native'
import { Button, List, Headline, Provider, useTheme, TextInput } from 'react-native-paper'
import ViewSong from './ViewSong'

/**
 * Query structure for query to backend
 * Used in useQuery call in DBList component
 */
export const GET_FILTERED_SONGS = gql`
query GetFilterSongs(
    $filter: String!
    $sortField: String
    $limit: Int
    $offset: Int
    ) {
        filterSongs(filter: $filter, sortField: $sortField, limit: $limit, offset: $offset){
            _id
            title
            duration
            album
            year_released
            artist{
                name1
                name2
            }
        }}
        `


/**
 * Exports DBList.tsx component to ToggleView.tsx
 * Component for Search, Sort, display DB as List, hosts detail-view of songs
 * @returns JSX elements for DBList
 */
const DBList = () => {
    const { colors } = useTheme();
    const theme = useTheme();
    const [searchQuery, setSearchQuery] = React.useState('');

    /** Functionality for: VIEWSONG
     * State for opening and closing ViewSong-modal
     * Stores data about the song that was clicked on in state property (sends property to ViewSong component further down the file)
     */
    const [property, setProperty] = useState({ _id: "", title: "", duration: 0, album: "", year: 0, artist: [""] })
    function getView(_id: string, title: string, duration: number, album: string, year: number, artist: Array<any>) {
        let artistArray: any[] = []
        artist.map((name: any) => {
            artistArray.push(name.name1)
            if (artist[0].name2) {
                artistArray.push(", ", name.name2)
            }
        })
        setProperty({ _id: _id, title: title, duration: duration, album: album, year: year, artist: artistArray })
    }
    /** State and functionality for showing/hiding viewSong-modal */
    const [visible, setVisible] = React.useState(false);
    const showModal = () => setVisible(true)


    /** Functionality for: SEARCH/FILTER AND LISTVIEW */
    const dispatch = useAppDispatch();
    /**
     * Variables for searchTerm. SearchTerm is used for saving values in searchbar onChange, and fetching query at load, while prevSearchTerm is used for fetching more queries.
     * PrevSearchTerm uses redux state searchTerm
     */
    let searchTerm = ""
    let prevSearchTerm2 = useAppSelector(selectSearchTerm);
    /**
     * Variables for sortTerm. SorTterm is used for saving values from chosen sort, and fetching query at load, while prevSortTerm is used for fetching more queries.
     * PrevSearchTerm uses redux state searchTerm
     */
    let sortTerm = useAppSelector(selectSortTerm)
    let savedSortTerm2 = useAppSelector(selectSortTerm);

    /**
     * Fetching the data from the query. 
     * Takes in parameters for filter, with field to sort by, the limit of how many elements should be fetched and the offset of which elements are fetched
     * Loading is displayed at the start of the fething of data, and if there is an error in the fetch, error is displayed
     */
    const { error, loading, data, fetchMore } = useQuery(GET_FILTERED_SONGS, {
        variables: {
            filter: searchTerm,
            sortField: sortTerm,
            limit: 10,
            offset: 0
        }
    })
    if (loading) return (
        <View style={styles.infoContainer}>
            <View style={styles.infoBox}>
                <Text>Loading...</Text>
            </View>
        </View>
    )
    if (error) return (
        <View style={styles.infoContainer}>
            <View style={styles.infoBox}>
                <Text>An error occured while loading.</Text>
                <Text>You might not be connected to the NTNU network or your VPN is off.</Text>
                <Text>Error: {error.message}</Text>
            </View>
        </View>
    )


    /** GET ARTIST
     * @param artist1 of song as string
     * @param artist2 of song as any (can be undefined or null, because it's not required)
     * @returns 
     */
    function getArtistString(artist1: String, artist2: any) {
        if (typeof (artist2) === "string") {
            let artists: String = artist1 + ", " + artist2
            return artists
        }
        return artist1
    }


    /** SORTING
     * Function for fetching more data, with the previous searchTerm (the one saved),
     * a limit with is determined by the length of the current amout of displayed data and the current sortTerm
     * @returns data, from a new query with new variables
     */
    function fetchingMoreDataSort() {
        dispatch(updateSortTerm(sortTerm))
        let offSetDataLength = parseInt(data.filterSongs.length)
        return fetchMore({
            variables: { filter: prevSearchTerm2, limit: offSetDataLength, sortField: sortTerm },
            updateQuery: (prevResult, { fetchMoreResult }) => {
                return fetchMoreResult;
            }
        })
    }


    /**
     * Functionality for search, fetches correct items from db
     * @param query from input
     */
    const onChangeSearch = (query: any) => {
        setSearchQuery(query)

        dispatch(updateSearchTerm(searchQuery))
        fetchMore({
            variables: { filter: searchQuery, sortField: sortTerm },
            updateQuery: (prevResult, { fetchMoreResult }) => {
                return fetchMoreResult
            }
        })
    }

    /**
     * Function for closing viewSong detail view.
     * Refreshes displayed list to remove deleted song
     */
    function onClose() {
        setVisible(false)
        fetchingMoreDataSort()
    }

    return (
        <>
            <Provider theme={theme}>
                <View style={styles.body}>
                    <ScrollView style={styles.scroll}>
                        <View >
                            <Headline style={styles.headline}>THE SONG COLLECTION</Headline>
                        </View>

                        {/* SEARCHBAR */}
                        <TextInput
                            placeholder="Search all songs"
                            onChangeText={onChangeSearch}
                            value={searchQuery}
                            accessibilityLabel={'Searchbar'}
                            style={styles.searchbar}
                            autoComplete="off"
                        />

                        {/* SORT BUTTONS */}
                        <View>
                            <Button mode="contained" style={styles.sortButtonStyle} onPress={() => { setSearchQuery("") }}>Clear input</Button>
                            <Button mode="contained" style={styles.sortButtonStyle} onPress={() => { sortTerm = 'title', fetchingMoreDataSort() }}>Sort by title (ASC)</Button>
                            <Button mode="contained" style={styles.sortButtonStyle} onPress={() => { sortTerm = "artist", fetchingMoreDataSort() }}>Sort by artist (ASC)</Button>
                            <Button mode="contained" style={styles.sortButtonStyle} onPress={() => { sortTerm = "", fetchingMoreDataSort() }}>Remove sort</Button>
                            <Button mode="contained" style={styles.sortButtonStyle} onPress={() => { onChangeSearch("") }}>Reset list</Button>
                        </View>

                        {/* LISTVIEW */}
                        <View style={styles.listBox}>
                            {data.filterSongs.map((song: { _id: string; title: string; duration: number; album: string; year_released: number, artist: Array<any> }) => {
                                return (
                                    <List.Item
                                        key={song._id}
                                        style={styles.listItemStyle}
                                        titleStyle={{ color: colors.text }}
                                        descriptionStyle={{ color: colors.text }}
                                        accessibilityLabel={'Song:' + song.title}
                                        title={song.title}
                                        titleEllipsizeMode="tail"
                                        description={
                                            song.artist.map((artist: { name1: String; name2: String | null }) => (
                                                <>
                                                    <View>
                                                        <Text>
                                                            {getArtistString(artist.name1, artist.name2)}
                                                        </Text>
                                                    </View>
                                                </>
                                            ))}
                                        descriptionNumberOfLines={3}
                                        descriptionEllipsizeMode="tail"
                                        data-cy="listItem"
                                        onPress={() => {
                                            showModal()
                                            getView(song._id, song.title, song.duration, song.album, song.year_released, song.artist)
                                        }}
                                    />
                                )
                            })}
                        </View>

                        {/* LOAD MORE-BUTTON */}
                        <View style={styles.buttonsBoxCenter}>
                            <Button data-cy="showMore" style={styles.buttons} mode='contained' onPress={() => {
                                let offSetDataLength = parseInt(data.filterSongs.length)
                                fetchMore({
                                    variables: { filter: prevSearchTerm2, offset: offSetDataLength, sortField: savedSortTerm2 },
                                    updateQuery: (prevResult, { fetchMoreResult }) => {
                                        fetchMoreResult.filterSongs = [
                                            ...prevResult.filterSongs,
                                            ...fetchMoreResult.filterSongs
                                        ];
                                        return fetchMoreResult;
                                    }
                                })
                            }}>Load more</Button>
                        </View>
                    </ScrollView>


                    {/* VIEWSONG */}
                    {visible &&
                        <>
                            <TouchableOpacity style={styles.modalBackground} onPress={() => { onClose() }}></TouchableOpacity>
                            <View style={styles.modalStyle} >
                                <ScrollView style={styles.modalScroll}>
                                    <Button mode='contained' onPress={() => { onClose() }}>Close</Button>
                                    <ViewSong param={property} />
                                </ScrollView>
                            </View>
                        </>
                    }
                </View>
            </Provider>
        </>
    )
}



export default DBList;

// S T Y L I N G :

const screenHeight = Platform.OS === "web" ? Dimensions.get('screen').height + 20 : Dimensions.get('screen').height
const windowHeight = Platform.OS === "web" ? Dimensions.get('window').height + 20 : Dimensions.get('window').height

const styles = StyleSheet.create({
    body: {
        width: '100%',
        height: windowHeight - 150,
        justifyContent: 'center',
    },

    // ON LOAD AND ERROR
    infoContainer: {
        display: "flex",
        alignItems: "center",
        marginTop: 50,
    },
    infoBox: {
        backgroundColor: "#F0FAF1",
        borderBottomColor: "#a9a9a9",
        borderBottomWidth: 1,
        padding: 5,
    },

    // HEADER
    headline: {
        fontSize: 30,
        fontWeight: 'bold',
        marginBottom: Platform.OS === "web" ? 25 : 15,
        marginTop: Platform.OS === "web" ? 20 : 10,
        alignSelf: 'center',
    },

    //SEARCHBAR
    searchbar: {
        marginLeft: "5%",
        marginRight: "5%",
        marginBottom: 15,
        backgroundColor: "#F0FAF1",
        shadowColor: '#0c1618',
    },

    // SCROLL
    scroll: {
        height: screenHeight / 3,
        width: '100%',
    },

    // BUTTONS (Load more)
    buttons: {
        width: "50%",
        maxWidth: 500,
        display: "flex",
        marginBottom: 20,
        height: 40,
        justifyContent: 'center',
    },
    buttonsBoxCenter: {
        display: "flex",
        flexDirection: "row",
        justifyContent: "center",
        alignItems: 'center',
    },

    // SORT-BUTTONS
    sortButtonStyle: {
        marginLeft: "5%",
        marginRight: "5%",
        height: 45,
        marginBottom: 5,
        fontSize: 12,
        justifyContent: 'center'
    },

    // LIST
    listBox: {
        paddingVertical: 10
    },
    listItemStyle: {
        marginLeft: "5%",
        marginRight: "5%",
        marginBottom: 15,
        backgroundColor: "#F0FAF1",
        borderWidth: 1,
        borderRadius: 5,
        borderColor: "#004643",
        borderTopEndRadius: 20,
        borderBottomEndRadius: 20,
        borderBottomStartRadius: 20,
        paddingLeft: 30,
        shadowColor: '#0c1618',
        shadowOffset: { width: 3, height: 3 },
        shadowOpacity: 0.3,
        shadowRadius: 5,
        width: "90%",
    },

    // MODAL
    modalStyle: {
        backgroundColor: "#ecf9ed",
        borderColor: "black",
        borderWidth: 2,
        padding: 20,
        borderRadius: 4,
        width: "80%",
        marginLeft: Platform.OS === "web" ? "15%" : "10%",
        marginRight: Platform.OS === "web" ? "15%" : "10%",
        maxHeight: Platform.OS === "web" ? screenHeight / 1.5 : screenHeight - 180,
        maxWidth: Platform.OS === "web" ? "70%" : "80%",
        marginTop: 0,
        position: 'absolute',
        display: "flex",
        alignItems: 'center',
        zIndex: 4,
    },
    modalBackground: {
        backgroundColor: "#c1ebc5",
        opacity: .7,
        height: Platform.OS === "web" ? screenHeight - 299 : screenHeight - 150,
        marginTop: Platform.OS === "web" ? 0 : 0,
        top: 0,
        zIndex: 3,
    },
    modalScroll: {
        width: "100%",
    },
})