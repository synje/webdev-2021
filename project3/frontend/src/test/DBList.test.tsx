import React from "react";
import { MockedProvider } from "@apollo/react-testing";
import { ReactTestRendererJSON } from 'react-test-renderer'
import TestRenderer from 'react-test-renderer';
import DBList, { GET_FILTERED_SONGS } from "../components/DBList";
import { store } from '../app/store';
import { Provider } from "react-redux";
import { render, waitForElementToBeRemoved } from "@testing-library/react";
import fireEvent from '@testing-library/user-event'

const mockFilterSong = {

    request: {
        query: GET_FILTERED_SONGS,
        variables: {
            filter: "",
            sortField: "",
            limit: 10,
            offset: 0
        },
    },
    result: {
        data: {
            filterSongs: [{
                _id: "1234",
                title: "Sang",
                duration: 1.23,
                album: "album",
                year_released: 2021,
                artist: [{
                    name1: "artist",
                    name2: undefined
                }]
            }]
        }
    }
};

test('Snapshot DBList', async () => {

    const component = TestRenderer.create(
        <Provider store={store}>
            <MockedProvider mocks={[mockFilterSong as any]}>
                <DBList />
            </MockedProvider>
        </Provider>
    )

    const { asFragment, getByText } = render(
        <Provider store={store}>
            <MockedProvider mocks={[mockFilterSong as any]}>
                <DBList />
            </MockedProvider>
        </Provider>)

    await waitForElementToBeRemoved(() => getByText('Loading...'))


    expect(asFragment()).toMatchSnapshot()

})

it('DBList loades', async () => {
    const component = TestRenderer.create(
        <Provider store={store}>
            <MockedProvider mocks={[mockFilterSong as any]}>
                <DBList />
            </MockedProvider>
        </Provider>
    )

    const tree = component.toJSON() as ReactTestRendererJSON
    if (tree != null) {
        expect(tree.children).toContain('Loading...')

    }
}
)


test('Test loaded filterSong', async () => {
    const { getByText } = render(
        <Provider store={store}>
            <MockedProvider mocks={[mockFilterSong as any]}>
                <DBList />
            </MockedProvider>
        </Provider>)

    await waitForElementToBeRemoved(() => getByText('Loading...'))

    const textElement = getByText('Sang')
    expect(textElement).toBeInTheDocument()

})

test('Test redux', async () => {
    const { getByText, getByPlaceholderText } = render(
        <Provider store={store}>
            <MockedProvider mocks={[mockFilterSong as any]}>
                <DBList />
            </MockedProvider>
        </Provider>)

    await waitForElementToBeRemoved(() => getByText('Loading...'))

    const searchInput = await getByPlaceholderText('Search...')
    const searchButton = getByText('SEARCH')
    expect(searchButton).toBeInTheDocument()

    expect(searchInput.innerHTML).toBe("")
    const termInput = "Sel"
    fireEvent.type(searchInput, termInput)
    //fireEvent.click(searchButton) //gives error, but filter (searchTerm) is updated according to store value
})