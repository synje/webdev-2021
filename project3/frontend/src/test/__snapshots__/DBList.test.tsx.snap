// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot DBList 1`] = `
<DocumentFragment>
  <div
    class="css-8atqhb"
  >
    <div
      class="css-ldo4d5"
    >
      <span
        class="css-1b0m27m"
      >
        Search for songs, albums or artists
      </span>
    </div>
    <div
      class="css-1l4w6pd"
    >
      <div
        class="searchBar css-ldo4d5"
      >
        <div
          class="chakra-form-control css-1z0y902"
          role="group"
        >
          <label
            class="chakra-form__label css-1v3sb5r"
            for="searchInput"
            id="field-3394560505-4-label"
          >
            Search
          </label>
          <input
            class="chakra-input searchBar css-0"
            data-cy="searchBar"
            id="field-3394560505-4"
            name="searchInput"
            placeholder="Search..."
            type="text"
          />
          <button
            class="chakra-button submitbtn css-1y2wz87"
            data-cy="searchButton"
            type="submit"
          >
            SEARCH
          </button>
        </div>
      </div>
    </div>
    <div
      class="sortingNav css-y6q02b"
    >
      <div
        class="css-7822g1"
      >
        <button
          class="chakra-button css-ja44gt"
          data-cy="sortbyTitle"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Sort by title (ASC)
        </button>
        <button
          class="chakra-button css-ja44gt"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Sort by duration (ASC)
        </button>
        <button
          class="chakra-button css-ja44gt"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Sort by album (ASC)
        </button>
        <button
          class="chakra-button css-ja44gt"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Sort by year released (ASC)
        </button>
        <button
          class="chakra-button css-ja44gt"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Sort by artist (ASC)
        </button>
        <button
          class="chakra-button css-ja44gt"
          data-cy="removeSort"
          style="background-color: floralwhite; border: 0.2px solid darkred; border-radius: 0; margin-right: 2px; margin-bottom: 2px;"
          type="button"
        >
          Remove sort
        </button>
      </div>
    </div>
  </div>
  <div
    class="css-ldo4d5"
  >
    <ul
      class="css-1875gv5"
      role="list"
    >
      <li
        class="listItems css-1lt61k4"
        data-cy="listItem"
        style="background-color: darkred;"
      >
         
        <p
          class="chakra-text listText css-j5b9iu"
          id="songCount"
          style="color: white; font-size: 17px;"
        >
          1
        </p>
        <p
          class="chakra-text listText css-2j9m6m"
          id="songTitle"
          style="color: white; font-size: 17px;"
        >
          Sang
        </p>
        <p
          class="chakra-text listText css-2j9m6m"
          id="songArtist"
          style="color: white; font-size: 17px;"
        >
          artist
        </p>
        <button
          class="chakra-button css-hek6jo"
          data-cy="songView"
          type="button"
        >
          VIEW SONG
        </button>
      </li>
    </ul>
  </div>
  <div
    class="css-3rg340"
  >
    <button
      class="chakra-button css-gocx0i"
      data-cy="showMore"
      style="background-color: floralwhite; border: 0.2px solid dimgray;"
      type="button"
    >
      Show more
    </button>
    <button
      class="chakra-button css-5c2176"
      data-cy="showLess"
      style="background-color: floralwhite; border: 0.2px solid dimgray;"
      type="button"
    >
      Show less
    </button>
  </div>
</DocumentFragment>
`;
